internal/provider/resource_vault_secrets_app.go

```diff
@@ -0,0 +1,169 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	clients "github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+func NewVaultSecretsAppResource() *resourceVaultsecretsApp {
+	return &resourceVaultsecretsApp{}
+}
+
+type resourceVaultsecretsApp struct {
+	client *clients.Client
+}
+
+func (r *resourceVaultsecretsApp) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_vault_secrets_app"
+}
+
+func (r *resourceVaultsecretsApp) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			// TODO: Add validators
+			"app_name": schema.StringAttribute{
+				Required: true,
+			},
+			"description": schema.StringAttribute{
+				Optional: true,
+			},
+		},
+	}
+}
+
+func (r *resourceVaultsecretsApp) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	r.client = client
+}
+
+type App struct {
+	ID          types.String `tfsdk:"id"`
+	AppName     string       `tfsdk:"app_name"`
+	Description string       `tfsdk:"description"`
+}
+
+func (r *resourceVaultsecretsApp) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan App
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: r.client.Config.OrganizationID,
+		ProjectID:      r.client.Config.ProjectID,
+	}
+
+	res, err := clients.CreateVaultSecretsApp(ctx, r.client, loc, plan.AppName, plan.Description)
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "Cannot create app")
+		return
+	}
+
+	plan.ID = types.StringValue(res.Name)
+	plan.AppName = res.Name
+	plan.Description = res.Description
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
```

Comment:
nit: you'll commonly see the setting of state at the end of a CRUD function reduced to one line
```go
resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
```